### DISCUSSION TITLE
Add support for British imperial volume units

### DISCUSSION THREAD

**Discussion by epenet** [home-assistant/architecture#813] [Entity Models]
Currently, so-called imperial units all refer to their US value, which are not always equal to their UK counterpart.
I am also not sure that US units should really be called "imperial" as those more often refer to the UK empire.

> **Not only are the number of ounces in pints, quarts, and gallons all larger in the imperial system, the size of one fluid ounce is also different**

For example:
- `1 * us fl oz` is equal to `28.41 mL`
- `1 * imperial fl oz` is equal to `29.57 mL`
- `1 * us fl oz` is equal to `0.96076 * imperial fl oz`

I suggest that the two current unit constants are deprecated and replaced with US versus UK units:
- VOLUME_GALLONS: Final = "gal"
- VOLUME_FLUID_OUNCE: Final = "fl. oz."

- VOLUME_US_GALLONS: Final = "gal (US)"
- VOLUME_US_FLUID_OUNCE: Final = "fl. oz. (US)"
- VOLUME_IMPERIAL_GALLONS: Final = "gal (imperial)"
- VOLUME_IMPERIAL_FLUID_OUNCE: Final = "fl. oz. (imperial)"

Linked discussions:
- automatic conversion: #809
- custom unit systems: #814

For reference, these are the conversion rules from pint:
```
@group USCSLiquidVolume
    minim = pint / 7680
    fluid_dram = pint / 128 = fldr = fluidram = US_fluid_dram = US_liquid_dram
    fluid_ounce = pint / 16 = floz = US_fluid_ounce = US_liquid_ounce
    gill = pint / 4 = gi = liquid_gill = US_liquid_gill
    pint = quart / 2 = pt = liquid_pint = US_pint
    fifth = gallon / 5 = _ = US_liquid_fifth
    quart = gallon / 4 = qt = liquid_quart = US_liquid_quart
    gallon = 231 * cubic_inch = gal = liquid_gallon = US_liquid_gallon
@end

@group ImperialVolume
    imperial_minim = imperial_fluid_ounce / 480
    imperial_fluid_scruple = imperial_fluid_ounce / 24
    imperial_fluid_drachm = imperial_fluid_ounce / 8 = imperial_fldr = imperial_fluid_dram
    imperial_fluid_ounce = imperial_pint / 20 = imperial_floz = UK_fluid_ounce
    imperial_gill = imperial_pint / 4 = imperial_gi = UK_gill
    imperial_cup = imperial_pint / 2 = imperial_cp = UK_cup
    imperial_pint = imperial_gallon / 8 = imperial_pt = UK_pint
    imperial_quart = imperial_gallon / 4 = imperial_qt = UK_quart
    imperial_gallon = 4.54609 * liter = imperial_gal = UK_gallon
    imperial_peck = 2 * imperial_gallon = imperial_pk = UK_pk
    imperial_bushel = 8 * imperial_gallon = imperial_bu = UK_bushel
    imperial_barrel = 36 * imperial_gallon = imperial_bbl = UK_bbl
@end
```

**Comment by ryanm101** [comment]
I'd widen this to support Hybrid Measurements as other countries such as Canada also don't use a single system exclusively.

**Reply by epenet** [reply]
This discussion is not about automatic conversions (https://github.com/home-assistant/architecture/discussions/809) and/or new unit systems (https://github.com/home-assistant/architecture/discussions/814).
The objective of this discussion is only to discuss which individual units are missing on the current "per-sensor" override.

I have adjusted the description accordingly

**Comment by TomBrien** [comment]
I've been thinking about this a bit more. Do American's refer to their unit system as "imperial". I (British bias) consider "imperial" to mean British and similar. Most references online seem to support this (with US units called US Customary Units). Should we rename  our current "imperial" to "US" or something else?

- https://en.m.wikipedia.org/wiki/United_States_customary_units
- https://en.m.wikipedia.org/wiki/Imperial_units
- https://www.google.com/amp/s/blog.ansi.org/2018/06/us-customary-system-history-units/%3Famp%3D1

**Reply by steverep** [reply]
I'm from the US and also an engineer.  The average person isn't going to be able to tell you what SI or Imperial units are, they'll know what metric is, and they'll take a half right guess at what US Customary means.

In engineering, you'll find a distinction between SI and Imperial/US/English/Standard, but those latter terms are often used interchangeably.

**Reply by epenet** [reply]
Because Imperial and US differ only on volume, I can definitely imagine that outside of volume they would be used interchangeably.
Regarding the naming of the "unit system", I think that "US Customary" would make sense, but if decide to deprecate the whole UnitSystem then I don't think it's worth spending time on it.

**Reply by TomBrien** [reply]
My concern is the use of imperial units in Canada and thus potentially more users there than in the UK but really need someone Canadian to confirm.

**Comment by epenet** [comment]
Deprecating VOLUME_GALLONS and VOLUME_FLUID_OUNCE would be a big breaking change.
I have adjusted the core PR https://github.com/home-assistant/core/pull/79580 to simply add two new units:

- `VOLUME_UK_GALLONS: Final = "gal (UK)"`
- `VOLUME_UK_FLUID_OUNCE: Final = "fl. oz. (UK)"`

I am still unsure about the naming: "UK" or "Imperial"?

**Reply by ryanm101** [reply]
Either/Or is fine. We'd refer to them as Imperial here but I'm happy to see a string say `gal (UK)` and it's arguably less confusing for non-technical people and my guess is most would be fine with that given the origin of the units was UK. 
I'm speaking as a UK native so I don't know if i said imperial to a Canadian for example if they would think UK or US, but the UK/US designator on the string would clarify that for anyone i would think and if they were still confused they could just google what they use / what is the difference.

**Reply by steverep** [reply]
I don't think using "UK" to distinguish them is a good idea.  As @TomBrien pointed out, there are quasi-official terms used commonly in references and legislation:
1. SI
2. Imperial
3. US customary

I think sticking with those terms is the correct approach, even if it means relabeling existing units.

**Reply by epenet** [reply]
I have adjusted https://github.com/home-assistant/core/pull/79580 to use "Imperial" terminology for the new units.

**Reply by frenck** [reply]
Just noticed the PR, and came here. This is not how this should be solved.
We are basically butchering the unit of measurement to solve a different issue. If anything the units of measurement used should be as clean as possible. Hacking our way around it, will bit our asses for the long future ahead.

**Reply by epenet** [reply]
Any thoughts on how to solve this?
Maybe VOLUME entities should be reporting an extra attribute to indicate "unit of measurement flavour" (USCS / Imperial), which default to USCS ?

**Reply by frenck** [reply]
> Any thoughts on how to solve this?

Sure, don't mix up units and systems.

**Reply by TomBrien** [reply]
#My feeling would be to rename the current "imperial" system to "US" maybe with the "customary" label on the end and create a _new_ "imperial" system. All users that had had "imperial" selected are moved to "US" and can then pick "imperial" if preferred. Flag it as a breaking change (even though it probably isn't) just to make it clear

**Reply by steverep** [reply]
I didn't really examine the PRs, but yes it should be solved by having 3 units systems.  If volume conversions currently use US Customary conversions, then I agree to rename it and add a 3rd for Imperial.

**Reply by KevinCathcart** [reply]
I've literally never seen any software that deals with unit conversions treat the Imperial units and US units (for the handful of units in which they meaningfully differ, mostly tons/hundredweights, and liquid volume measurements) as different interpretations of the same unit. I've always seen them treated as two separate units. This includes everything from libraries like `pint`, to the linux `units` command, to the Windows calculator. 

Thus I disagree with Frenck about Imperials Gallons and US Gallons being the same unit (just different interpretations based on the unit system in use), or similarly with Fluid Ounces.

Personally, I see them as completely different units whose symbol and display name may vary as part of localization (either by language region tag "en-US vs en-UK", or by some preferred units setting.)  The alternative is to require that units that may have more than one interpretation always be accompanied by a unit system. This is because without that information performing accurate unit conversion is impossible.

What do I mean by differing localization?

I mean that "Gallon (Imperial)" might show up as just "Gallon" to people in the UK (Who have their settings correct), but should show as "Imperial Gallon" or "Gallon (Imperial)", to somebody in the US ( who have their setting correct). To people in regions that never really used imperial units, it is probably always important to be specific as to which version of the gallon is being referenced (otherwise it would be very easy to confuse the two).   Localizing units like that is nothing really new, since units already need to be localized: "meter" is "mètre" in French, "day" and "year" vary by language, etc.

As far as correct settings, there are a few ways of handling that localization. One is to handle it based on unit system selected. So selecting the US unit system would have US Gallons show as just "Gallons", while Imperial gallons would show as "Gallons (Imperial)". Another option is to go based on language code. With "en-US" users seeing US gallons as "gallons", while "en-UK" users would see imperial gallons as "Gallons".

Example for why paring value with unit system would always be necessary if treated as the same unit:

Somebody in the UK would not find it acceptable to have a device they imported from somewhere that reports volume in (US) gallons have that figure displayed without conversion on screen as "Gallons", as that would imply Imperial Gallons to that user. If it was displayed as "Gallons (US)", that would be less confusing, although the user would probably prefer to see either Imperial Gallons or Liters. (Unless they are American expats, in which case, who knows what they would prefer?) If the user elected to display it in Liters, it would be important that the conversion be done using the US Gallon to Liter ratio, (since that is how the device measures). You would get a wrong result using the Imperial Gallon to Liter conversion ratio. Thus any value tagged as gallons must always have information about which type of gallon too.

**Reply by steverep** [reply]
I agree with you except for displaying the units differently depending on your location or language selection.  They should display the same way or undoubtedly confusion will occur.

**Reply by KevinCathcart** [reply]
The changing of display would be what users would probably expect. Users in the US would normally expect to see "Gallons" not "Gallons (US)". However if some other type of gallon was being used, they would certainly expect it to be shown.

That said, I see no harm in always showing the suffix, except that it seems like it could plausibly likely to lead to user complaints. That said, These units are not especially common ones for Home Assistant (which is why we have gotten away with not distinguishing the Imperial system from the US Customary system so far).

**Reply by frenck** [reply]
I do see harm in always showing the suffix, that was my whole point in being against it.

You will not find any suffix on a can of milk in either the US or the UK. The point being, it's not the unit that need suffixing... it might need extra context, but the name of the unit doesn't change.

It is user facing, ugly, and above all, not an real unit by its own anymore (making it incompatible with libraries like pint as you mentioned earlier).

**Reply by steverep** [reply]
@frenck I don't think it'd be as confusing as you suspect and HA is much more complicated than a jug of milk.  If an uninformed person saw "gal (US)" or "gal", they certainly would think "gallons" for both and move on.  But if an informed person sees "gal", they'll wonder "US or Imperial?", which is why this discussion started.

FWIW, common unit converters like Windows calculator or Google all distinguish them and never show just "gal".

Personally, I would concede to show just "gal" only on the condition that the unit system in your profile - SI, Imperial, or US Customary - is fixed and forced upon each entity, so you know what you're getting when you see "gal".  But I see that as too restrictive and it's common in many countries and industries to mix unit systems depending on the application.  I should be able to choose whatever unit I want for an entity in any system.

**Reply by frenck** [reply]
> I don't think it'd be as confusing as you suspect and HA is much more complicated than a jug of milk. If an uninformed person saw "gal (US)" or "gal", they certainly would think "gallons" for both and move on. But if an informed person sees "gal", they'll wonder "US or Imperial?", which is why this discussion started.

I think I am not coming across, I am not saying we shouldn't make it clear... I am saying we should not solve it by messing with the unit name. Sure! In the ui we need to provide that context (when fitting in some places). For example, on a dashboard, maybe not showing it... as it clutters. In the more info dialog... yeah sure.

All I am saying is: we should not just smack it into the existing unit properly

Yes, we need to provide context, but context can be provided by using an extra property field, eg by providing a unit system property next to it. Or a variant property... of whatever...
Some else that provides that context for the unit.

**Reply by parautenbach** [reply]
I'm only following the conversation out of interest (with my engineering background). I have no preference among these alternative and/or old unit systems (I just go SI/ISO/metric whenever I can), but I thought I'd add this example into the mix. 

<img width="685" alt="Screenshot 2022-10-13 at 07 49 20" src="https://user-images.githubusercontent.com/743320/195512778-d94302d3-dbcc-4f19-a12a-aa2d1adf431f.png">

Maybe what's needed isn't only the unit (that should be short, as per frenck's milk example), but also a unit name, and the option of using the latter for display purposes.

**Reply by epenet** [reply]
I love that your example has two "Fluid ounce", but no US or Imperial flavour of it...

**Reply by KevinCathcart** [reply]
Frenck, yes, a two peice of Metadata (e.g. two attributes) solution (unit name and variation) absolutely can work, no question about it.

From my point of view it seems like the more complicated approach that adds a bunch of extra edge cases. 

Let's take a look at some of those:

The first what do we do if an integration provides a state with a unit attribute for an ambiguous unit but fails to provide a variant property. That could perhaps "work" if no unit conversion ever gets done on the value, but as soon as you need to convert the value, you really need to know the variant since different variants will have different conversion factors. Perhaps you could "guess" that it matches the unit system specified by the user, but if that guess is wrong, you get a bad conversion, and if the user specified metric as the unit system that does not provide any help in guessing the conversion factor to use.

The second is that if somebody who has their unit system marked as metric, and a sensor that reports in Liters, and they want to use the UI feature where they can chose some other unit to display the value in (with automatic conversion), what happens if they select "gallon"? In that scenario, the user's unit system preference is no help, since we just said it was set to metric, so now we need to ask them which variant of gallon to use. Certainly possible to do, but a bit of an extra complication vs a menu with "Gallon (US)" and "Gallon (Imperial)" as options alongside unambiguous units like "liter".

Then there is the question for non-ambigous units. Do we ban the variant attribute? Is it allowed but optional? What if it is an absurd combination like "meter" and "Imperial"? Meter is not part of the Imperial system, but is part of the Metric and SI systems. 

By having a single metadata field that combines both the unit and the variant (but only when needed), you can sidestep all of those, at the cost of possibly wanting to be fancy on the front end to strip out the variant information when users would not be expecting it.

**Reply by frenck** [reply]
Unit are part of a unit system. If entries only provide units, it will never be clear which system they are in.

Prefixing/postfixing is not a way to solve that. Adding meta data on which unit system it is in, does solve that.

Yes, if something advertises the unit to be meter, is should also advertise what the unit system is. It could be metric, it could be SI.

If it is gallon an in the US unit system, than we know how to calculate with it and can decide how we display it.

Additionally, giving this meta data, allow to enforce checks in our state engine (is the unit system provided? Does this unit exists in the unit system?)

**Reply by ryanm101** [reply]
There is one other solution though i'm not sure it is a good one.. 
If all integrations are Forced/Mandated to convert the device data to a common unit (lets say metric for the purpose of discussion). Then all the logic for display is pushed to the UI to convert based on what the user wants to see. 

The advantages here are:
*  HA as a system does not care what the device actually returns as it is converted by the integration before ingestion, all calculations in HA are done on a single unified system which makes comparisons easy for automations etc. 
* The 'average' user does not have to care about the units except what they see.
* Advanced users that are pulling raw data into influx only have one unit system to use and they can convert as they see fit.
* All Unit conversions (to metric) are the problem of the integration provider. (I suspect *most* devices will return metric by default from an engineering standpoint)

**Reply by KevinCathcart** [reply]
Frenck, if your idea is that basically every integration that exposes values with a unit should be required to add another attribute to specify a unit system, that seems excessive. It is not technically *wrong*, but it would imply the need to change a huge percentage of all integrations both core and custom, all to handle the tiny handful of ambiguous units.

The overwhelming majority of units have universal agreement these days. The meter as defined by SI is what everybody uses for the meter. The same with the kilogram,the Ampere, the Volt, etc. So each of those those are equivalent units (or most would say “the same unit”) in the Metric system, SI, the modern US system, the modern UK system, etc.

Requiring integrations to need to chose if they are providing a metric meter or an SI meter when those are literally the exact same thing, feels absurd to me. How can an integration author know if their device provides values in metric Liters, or SI Liters, given that those are by definition completely numerically identical. Look in the manual and use whichever word the manual author happened to arbitrarily pick? If the manual fails to specify, are they expected to call up the IOT product vendor to ask them?

Do you really want to force your fellow core devs to need to maintain some form of mapping of which units are included in the various unit systems? I suppose you could always try to defer that mapping to a library like `pint`, but do note that that particular library will happily tell you that the SI system contains a unit called “gallon”, (`ureg.sys.SI.gallon`), which happens to just be the US gallon. And obviously the SI system does not contain such a unit, nor is such a unit in the list of non-SI units approved to be used with SI.

**Reply by steverep** [reply]
@frenck you're right that I wasn't fully understanding your point, but I think I do now and I'm not sure I agree...

> Unit are part of a unit system. If entries only provide units, it will never be clear which system they are in.

Why is that a big deal? Yes, units are standardized in systems, but they stand on their own by their definitions in those systems.  And when two systems decide to clobber things by overlapping their unit names, the world responds by adjusting their names for clarity.  The system should be used as nothing more than a display preference for users, which may often break BTW because of industry common use (without most users even noticing).  For example, a light bulbs power should be displayed in Watts and temperature in Kelvin even if my preference is US, because no one uses horsepower and Rankine for those things.

> Yes, if something advertises the unit to be meter, is should also advertise what the unit system is. It could be metric, it could be SI.

This just unnecessarily forces everyone to become an expert on unit systems and standards.  Units for time and electricity are good examples.  The world measures time units in seconds and electric current in amperes, both defined in SI and there is no equivalent Imperial or US unit.  Most people don't know that and don't need to.

> Additionally, giving this meta data, allow to enforce checks in our state engine (is the unit system provided? Does this unit exists in the unit system?)

Now you're signing up to keep up with legislation and standards bodies that might cross-include units in the future.  I don't recommend that.

**Reply by epenet** [reply]
I was just looking at temperature units (`°C` and `°F`) and if we go all the way then we should drop the system (`C` / `F`) to keep just the unit (`°`). But I don't think that would makes sense either!

I think that `core` should include the unit system inside the `unit_of_measurement` for all "confusing" units:
- `gal (US)`, `°F`
- `gal (UK)`, `°C`

If frontend is a concern (and in my opinion it is not but others seem to disagree) then it should be handled there to have a mapping for "display units":
- `°F` => `°`
- `°C` => `°`
- `gal (US)` => `gal`
- `gal (UK)` => `gal`

Or we could even make it cleverer to keep the unit system if it doesn't match the current unit system.
For example, on a metric system, displaying `°` for Celsius and displaying `°F` for Farenheit.

![image](https://user-images.githubusercontent.com/6771947/198003377-fb6cb481-8b82-4d1f-b0d7-d66725e9f9e7.png)

**Comment by thecode** [comment]
Apologize for the dirt on the screen, this is how my Skoda car let you set the unit system (there are also options for temperature, pressure  and time):
![image](https://user-images.githubusercontent.com/1858925/196374513-bf15f4e5-8152-4b78-9e69-571ccfdeab29.png)
Mainly to show there are different settings for US/UK mpg, but there is also setting for different user preferences which is not related to a region or the measurement standard unit.

**Reply by ryanm101** [reply]
@thecode out of interest, if you connect ODBII to the car what 'UNIT' does it show? might be hard to see as likely the ODBII software will offer you a choice.

**Reply by thecode** [reply]
I will try to get a ODB-II adapter to check
